hibernate-core/src/main/java/org/hibernate/dialect/pagination/AbstractLimitHandler.java

```diff
@@ -178,4 +187,31 @@ protected final int getMaxOrLimit(RowSelection selection) {
 			return maxRows;
 		}
 	}
+
+	/**
+	 * The offset/limit clauses typically must come
+	 * before the {@code FOR UPDATE}ish clauses, so
+	 * we need a way to identify these clauses in
+	 * the text of the whole query.
+	 */
+	protected Pattern getForUpdatePattern() {
+		return FOR_UPDATE_PATTERN;
+	}
+
+	/**
+	 * Insert a fragment of SQL right before the
+	 * {@code FOR UPDATE}ish clauses at the end
+	 * of the query.
+	 */
+	protected String insertBeforeForUpdate(String limitOffsetClause, String sqlStatement) {
+		Matcher forUpdateMatcher = getForUpdatePattern().matcher( sqlStatement );
+		if ( forUpdateMatcher.find() ) {
+			return new StringBuilder( sqlStatement )
+					.insert( forUpdateMatcher.start(), limitOffsetClause )
```

Comment:
I think we need to. use `" " + limitOffsetClause` to avoid the bug that there would be no whitespace delimiter between the preceding text and 'limitOffsetClause'.